Fix plugin removal and clean code

ret2libc opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

- when removing a plugin through `L-` the "plugins" list in RzCore/RzDemangler/RzAsm/etc. was not updated. Define a _dt callback in the plugins handler for each type to make sure that list is updated
- clean the lib.c code and document it
- move dlopen/dlsym/dlclose to sys.c

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

- compile an external plugin
- load it
- check it works
- unload it
- check it is not present anymore

ret2libc committed:
cmd: Fix summary of few "L" sub-commands

ret2libc committed:
util: start decluttering plugins handling

- Add Doxygen docs
- Remove unused API functions and make others internal only
- Use RzLibType enum for plugin type instead of int

ret2libc committed:
util: move dlopen/dlsym/dlclose to rz_sys

ret2libc committed:
util: make rz_lib_open/close return bool

ret2libc committed:
util: simplify rz_lib_add_handler and rz_lib_open*

ret2libc committed:
libs: implement rz_*_plugin_del to remove a plugin

ret2libc committed:
core: remove commands from RzCmd when removing plugins

XVilka commented on librz/analysis/analysis.c:
```diff
@@ -177,11 +177,20 @@ RZ_API RzAnalysis *rz_analysis_free(RzAnalysis *a) {
 	return true;
 }
 
+RZ_API bool rz_analysis_plugin_del(RzAnalysis *analysis, RzAnalysisPlugin *p) {
```
Add `RZ_NONNULL` for `p`

XVilka commented on librz/asm/asm.c:
```diff
@@ -330,7 +330,7 @@ RZ_API void rz_asm_free(RzAsm *a) {
 }
 
-RZ_API bool rz_asm_add(RzAsm *a, RzAsmPlugin *p) {
+RZ_API bool rz_asm_plugin_add(RzAsm *a, RzAsmPlugin *p) {
```
Add `rz_return_if_fail(a & p, false)` and `RZ_NONNULL` for `p`

XVilka commented on librz/asm/asm.c:
```diff
@@ -341,9 +341,16 @@ RZ_API bool rz_asm_add(RzAsm *a, RzAsmPlugin *p) {
-RZ_API int rz_asm_del(RzAsm *a, const char *name) {
-	/* TODO: Implement rz_asm_del */
-	return false;
+RZ_API bool rz_asm_plugin_del(RzAsm *a, RzAsmPlugin *p) {
```
Same

XVilka commented on librz/bin/bin.c:
```diff
@@ -417,7 +417,21 @@ RZ_API bool rz_bin_plugin_add(RzBin *bin, RzBinPlugin *foo) {
 }
 
-RZ_API bool rz_bin_xtr_add(RzBin *bin, RzBinXtrPlugin *foo) {
+RZ_API bool rz_bin_plugin_del(RzBin *bin, RzBinPlugin *plugin) {
```
Same. Also in two previous it's `p` while here it's `plugin`, probably better to use the same name everywhere

XVilka commented on librz/bin/bin.c:
```diff
@@ -436,6 +450,20 @@ RZ_API bool rz_bin_xtr_add(RzBin *bin, RzBinXtrPlugin *foo) {
 	return true;
 }
 
+RZ_API bool rz_bin_xtr_plugin_del(RzBin *bin, RzBinXtrPlugin *plugin) {
```
Same

XVilka commented on librz/bp/bp_plugin.c:
```diff
@@ -3,7 +3,7 @@
 #include <rz_bp.h>
 
-RZ_API int rz_bp_plugin_del(RzBreakpoint *bp, const char *name) {
+RZ_API int rz_bp_plugin_del_byname(RzBreakpoint *bp, const char *name) {
```
I suggest to disallow `NULL` here and just do an assert

XVilka commented on librz/core/cplugin.c:
```diff
@@ -34,6 +34,11 @@ RZ_API bool rz_core_plugin_add(RzCore *core, RzCorePlugin *plugin) {
 	return true;
 }
 
+RZ_API bool rz_core_plugin_del(RzCore *core, RzCorePlugin *plugin) {
```
`RZ_NONNULL` for `plugin`

XVilka commented on librz/core/p/core_plugin_example.c:
```diff
@@ -116,6 +116,5 @@ _RZ_API RzLibStruct rizin_plugin = {
 	.type = RZ_LIB_TYPE_CORE,
 	.data = &rz_core_plugin_example,
 	.version = RZ_VERSION,
-	.pkgname = "example_package"
```
Why?

ret2libc replied:
It was used only at https://github.com/rizinorg/rizin/blob/dev/librz/util/lib.c#L350 and I'm not sure it was that useful. Maybe a `name` would be useful as a way to reference the plugins, but I'm unsure right now.

wargio replied:
i agree. we should remove this.

XVilka commented on librz/debug/plugin.c:
```diff
@@ -65,6 +65,16 @@ RZ_API bool rz_debug_plugin_add(RzDebug *dbg, RzDebugPlugin *foo) {
 	return true;
 }
 
+RZ_API bool rz_debug_plugin_del(RzDebug *dbg, RzDebugPlugin *plugin) {
```
`RZ_NONNULL`

XVilka commented on librz/egg/egg.c:
```diff
@@ -89,7 +89,7 @@ RZ_API RzEgg *rz_egg_new(void) {
 }
 
-RZ_API int rz_egg_add(RzEgg *a, RzEggPlugin *foo) {
+RZ_API int rz_egg_plugin_add(RzEgg *a, RzEggPlugin *foo) {
```
Same, probably also `plugin` instead of `foo`

XVilka commented on librz/egg/egg.c:
```diff
@@ -104,6 +104,11 @@ RZ_API int rz_egg_add(RzEgg *a, RzEggPlugin *foo) {
 	return true;
 }
 
+RZ_API int rz_egg_plugin_del(RzEgg *a, RzEggPlugin *plugin) {
```
`RZ_NONNULL`

XVilka commented on librz/io/io_plugin.c:
```diff
@@ -29,6 +29,21 @@ RZ_API bool rz_io_plugin_add(RzIO *io, RZ_BORROW RzIOPlugin *plugin) {
+	return true;
+}
+
+RZ_API bool rz_io_plugin_del(RzIO *io, RZ_BORROW RzIOPlugin *plugin) {
```
`RZ_NONNULL`

XVilka commented on librz/lang/lang.c:
```diff
@@ -112,7 +112,7 @@ RZ_API bool rz_lang_setup(RzLang *lang) {
 }
 
-RZ_API bool rz_lang_add(RzLang *lang, RzLangPlugin *foo) {
+RZ_API bool rz_lang_plugin_add(RzLang *lang, RzLangPlugin *foo) {
```
`RZ_NONNULL`, `plugin` instead of `foo`

XVilka commented on librz/lang/lang.c:
```diff
@@ -123,6 +123,11 @@ RZ_API bool rz_lang_add(RzLang *lang, RzLangPlugin *foo) {
 	return false;
 }
 
+RZ_API bool rz_lang_plugin_del(RzLang *lang, RzLangPlugin *plugin) {
```
Same

XVilka commented on librz/parse/parse.c:
```diff
@@ -42,7 +42,7 @@ RZ_API void rz_parse_free(RzParse *p) {
 }
 
-RZ_API bool rz_parse_add(RzParse *p, RzParsePlugin *foo) {
+RZ_API bool rz_parse_plugin_add(RzParse *p, RzParsePlugin *foo) {
```
Same, no `foo`, also `rz_return_if_fai(p & plugin, false)`

XVilka commented on librz/parse/parse.c:
```diff
@@ -53,6 +53,11 @@ RZ_API bool rz_parse_add(RzParse *p, RzParsePlugin *foo) {
 	return true;
 }
 
+RZ_API bool rz_parse_plugin_del(RzParse *p, RzParsePlugin *plugin) {
```
Same

XVilka commented on librz/util/lib.c:
```diff
@@ -7,172 +7,20 @@
+ * \param symnamefunc Name of the symbol pointing to a \p RzLibStructFunc function
+ * \return The new \p RzLib instance
+ */
 RZ_API RzLib *rz_lib_new(const char *symname, const char *symnamefunc) {
```
Add `RZ_NULLABLE`

XVilka commented on librz/util/lib.c:
```diff
@@ -330,67 +131,100 @@ static char *major_minor(const char *s) {
+ * \param file Dynamic library to load
+ * \return true if the plugin is correctly loaded, false otherwise
+ */
+RZ_API bool rz_lib_open(RzLib *lib, const char *file) {
```
Could `file` here be NULL? I suggest prohibiting such case

XVilka commented on librz/util/lib.c:
```diff
@@ -411,54 +245,53 @@ RZ_API bool rz_lib_opendir(RzLib *lib, const char *path, bool force) {
-			} else {
-				RZ_LOG_INFO("Cannot open %ls\n", dir.cFileName);
+		int cx = swprintf(file, _countof(file), L"%ls/%ls", wcpath, dir.cFileName);
+		if (cx >= 0) {
```
You could flip the condition and use error+`continue` in case of `< 0`

XVilka commented on librz/util/lib.c:
```diff
@@ -483,24 +320,25 @@ RZ_API bool rz_lib_opendir(RzLib *lib, const char *path, bool force) {
-	int type, const char *desc,
-	int (*cb)(RzLibPlugin *, void *, void *), /* constructor */
-	int (*dt)(RzLibPlugin *, void *, void *), /* destructor */
+	RzLibType type, const char *desc,
```
Could `desc` be `NULL`? If yes - please add `RZ_NULLABLE`, if not - an assert.

XVilka commented on librz/util/sys.c:
```diff
@@ -1949,3 +1949,72 @@ RZ_API bool rz_sys_stop(void) {
+/**
+ * \brief Implementation across systems to open a dynamic library
+ */
+RZ_API void *rz_sys_dlopen(const char *libname) {
```
`RZ_NULLABLE`

XVilka reviewed (changes requested):
The cleanup looks really good.

ret2libc committed:
librz: call plugins 'fini' method when removing a plugin

ret2libc committed:
core/cmd: do not fail if the hashmap removal fails

When removing a group command, the inner command might have the same
name, thus it would result in trying to remove the same name twice.
Ignore it, as it is fine.

ret2libc committed:
librz: make RzCrypto plugins handling like the others

ret2libc committed:
Use proper RZ_NONNULL, asserts and annotations

wargio reviewed (approved):
<3 nice cleanup

wargio commented:
something is extremely broken.

ret2libc committed:
librz: do not free plugins when removing from the list

Plugins are always defined as static structures in files, thus they
don't need to be deallocated.

ret2libc committed:
Use RZ_NONNULL also in RzCrypto plugins

XVilka commented:
Doesn't load proper crypto plugin for some reason:
```

[XX] db/cmd/cmd_wD woE woD aes-ecb
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 'w helloworld
woE aes-ecb s:0123456789abcdef
woD aes-ecb s:0123456789abcdef
ps
' =
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-helloworld
+;\x94\xd4\x0f\x8es\xec\xa6\x9e\x89\xe8F\x93\xaf4|\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5\x0b\x9b\x15\xdaKD\xa0\xf5\x15\x1d\xcf\xc4\xc0\x1f5\xd5

-- stderr
WARNING: core: Written 256 byte(s)
WARNING: core: Written 512 byte(s)
```

ret2libc committed:
librz/crypto: add rz_crypto_reset and use it

ret2libc committed:
librz: allocate the plugin structures before adding them

ret2libc committed:
Free plugins after the last use

XVilka reviewed (changes requested):
```
Mismatched function annotation for rz_io_plugin_add at <rizin/librz/include/rz_io.h:347:13> : RZ_NONNULL (was RZ_BORROW at <rizin/librz/io/io_plugin.c:13:13>
Mismatched function annotation for rz_io_plugin_add at <rizin/librz/include/rz_io.h:347:13> : RZ_BORROW (was None at <rizin/librz/io/io_plugin.c:13:13>
```

wargio commented on librz/include/rz_lib.h:
```diff
@@ -31,95 +29,111 @@ RZ_LIB_VERSION_HEADER(rz_lib);
 } RzLibPlugin;
 
-/* store list of initialized plugin handlers */
+typedef bool (*RzLibCallback)(RzLibPlugin *, void *, void *);
```
If the constructor has to return something, like a pointer, how can it? for the destructor this makes sense, but not for the constructor

ret2libc replied:
Where do you see the need to return anything here? I think the purpose is usually just to call something like `rz_crypto_plugin_add` and make sure RzCrypto is aware of the new plugin.

wargio replied:
the constructor should allow to set a user datastructure and the destructure to destroy it (we have the user pointer on purpose). but using this function declaration, it does not allow to set anything.

ret2libc replied:
The constructor can manipulate the `RzLibPlugin.data` pointer, which is usually `RzCryptoPlugin`, `RzAsmPlugin`, etc. and there it can manipulate everything. Do you have concrete examples of what would be useful here?

wargio commented on librz/include/rz_lib.h:
```diff
@@ -31,95 +29,111 @@ RZ_LIB_VERSION_HEADER(rz_lib);
-	int (*destructor)(RzLibPlugin *, void *user, void *data);
+	RzLibType type; ///< Plugin type this handler handles
+	char desc[128]; ///< Description of the handler
+	void *user; ///< User pointer
```
As said before, this is never going to be populated

ret2libc replied:
This is populated by rz_lib_add_handler and it is passed as the second argument of ->constructor. What is the problem there?

ret2libc committed:
Match signature of rz_io_plugin_add

wargio commented on librz/util/lib.c:
```diff
@@ -199,21 +52,19 @@ RZ_API void rz_lib_free(RzLib *lib) {
-	if (h && h->constructor) {
-		RZ_LOG_DEBUG("PLUGIN LOADED %p fcn %p\n", h, h->constructor);
+	if (h->constructor) {
 		return h->constructor(plugin, h->user, symbol->data);
```
this does not make sense. how a user can set the `h->user` pointer if you just pass a pointer (i.e. `void *`) instead of the reference to the pointer (i.e. `void**`) ?

ret2libc replied:
This is data to be passed to the plugin-type handler. Why do you need to set the user? I'm not understanding what you want to achieve exactly.

wargio replied:
ignore me.

ret2libc committed:
Merge branch 'dev' into removal-plugins
